docs/lang/articles/misc/ndarray_android.md

```diff
@@ -0,0 +1,449 @@
+vel = ti.Vector.ndarray(2, ti.f32, N)
+force = ti.Vector.ndarray(2, ti.f32, N)
+
+# Run the AOT module builder
+def aot():
+    m = ti.aot.Module(ti.opengl)
+    m.add_kernel(initialize, (pos, vel))
+    m.add_kernel(compute_force, (pos, vel, force))
+
+    dir_name = 'nbody_aot'
+    m.save(dir_name, '')
+aot()
+```
+
+**In line 3, you initialize Taichi:**
+
+1. Set `use_gles` to `True` to generate GLES compute shaders for Android.
+2. Set `allow_nv_shader_extension` to `False` to prevent the generated GLES compute shaders from using Nvidia GL extensions on Android.
+
+> This setting is because Android supports GLES APIs but GLES does not support `NV_SHADER_EXTENSION`.
+
+**In line 50-58, you define and build the Taichi AOT module:**
+
+1. Create a Taichi AOT module, specifying its backend as OpenGL:
+
+```python
+     m = ti.aot.Module(ti.opengl)
+```
+
+2. Add the required kernels `initialize` and `compute_force`, each with its own Ndarrays, to the module:
+
+```python
+m.add_kernel(initialize, (pos, vel))
+
+m.add_kernel(compute_force, (pos, vel, force))
+```
+
+3. Specify a folder under your current working directory for holding the files that the module generates:
+
+```python
+dir_name = 'nbody_aot'
+
+m.save(dir_name, '')
+```
+
+*The necessary compute shaders together with a JSON file appear under the specified directory.*
+
+### Parse the generated JSON file
+
+:::note
+
+From this section, you are required to come up with your own runtime program. We provide an [example Java runtime program for Android](https://github.com/taichi-dev/taichi-aot-demo/blob/master/nbody_ndarray/java_runtime/NbodyNdarray.java) for your reference, but you may need to adapt these codes for your platform and programming language.
+
+:::
+
+After generating the necessary GLES compute shaders, you need to write your runtime program to parse the following JSON file to some data structures. The JSON file contains all the necessary information for executing the compute shaders. Organized by Taichi kernel, it provides a clearer image of the compute shaders and Ndarrays in each kernel. Let's take a closer look at the structure.
+
+- **Organized by Taichi kernel**
+
+  - `initialize` (line 4)
+  - `compute_force` (line 51)
+
+- **Kernel-specific compute shaders**
+
+  Taking `initialize` as an example, the kernel has generated one compute shader named `initialize_c54_00` (line 7) and the other named `initialize_c54_01` (line 13).
+
+- **Kernel-specific** `args_buff`
+
+  The `initialize` kernel is assigned an `args_buffer` of `128` Bytes (line 21). Note that the size of `args_buffer` is dependent on the number of Ndarrays (`pos` and `vel`) that the kernel takes, (see `arg_count` in line 19). The `initialize` kernel, or each kernel more precisely, has a dedicated `args_buffer` for storing scalar arguments specified in `scalar_args` (line 27) and Ndarray shape information in accordance with what `array_args` (line 28-45) specifies.
```

Comment:
I'd personally prefer to remove line numbers here:
1. it's easier to maintain
2. we didn't show line numbers in the code snippet so users will have to count themselves, which is a bit difficult to use :P

Reply:
@rexwangcc has just added this feature to the site. You can see the line numbers when this PR is merged.